trustpinning/certs_test.go

```diff
@@ -203,11 +208,269 @@ func TestValidateRoot(t *testing.T) {
+
+	store, err := trustmanager.NewKeyFileStore(tempBaseDir, passphraseRetriever)
+	require.NoError(t, err)
+	cs := cryptoservice.NewCryptoService(store)
+
+	err = store.AddKey(trustmanager.KeyInfo{Role: data.CanonicalRootRole, Gun: "notary-signer"}, privKey)
+	require.NoError(t, err)
+
+	newTestSignedRoot, err := testRoot.ToSigned()
+	require.NoError(t, err)
+
+	err = signed.Sign(cs, newTestSignedRoot, newRootLeafKey)
+	require.NoError(t, err)
+
+	// Check that we validate correctly against a pinned CA and provided bundle
+	err = ValidateRoot(certStore, newTestSignedRoot, "notary-signer", TrustPinConfig{CA: map[string]string{"notary-signer": validCAFilepath}, DisableTOFU: true})
+	require.NoError(t, err)
+
+	// Add an expired CA for the same gun to our previous pinned bundle, ensure that we still validate correctly
+	certStore.RemoveAll()
+	goodRootCABundle, err := trustmanager.LoadCertBundleFromFile(validCAFilepath)
+	require.NoError(t, err)
+	memKeyStore := trustmanager.NewKeyMemoryStore(passphraseRetriever)
+	cryptoService := cryptoservice.NewCryptoService(memKeyStore)
+	testPubKey, err := cryptoService.Create("root", "notary-signer", data.ECDSAKey)
+	require.NoError(t, err)
+	testPrivKey, _, err := memKeyStore.GetKey(testPubKey.ID())
+	require.NoError(t, err)
+	expiredCert, err := cryptoservice.GenerateExpiredTestingCertificate(testPrivKey.CryptoSigner(), "notary-signer")
+	require.NoError(t, err)
+	bundleWithExpiredCert, err := trustmanager.CertChainToPEM(append(goodRootCABundle, expiredCert))
+	require.NoError(t, err)
+	bundleWithExpiredCertPath := filepath.Join(tempBaseDir, "bundle_with_expired_cert.pem")
+	require.NoError(t, ioutil.WriteFile(bundleWithExpiredCertPath, bundleWithExpiredCert, 0644))
+
+	// Check that we validate correctly against a pinned CA and provided bundle
+	err = ValidateRoot(certStore, newTestSignedRoot, "notary-signer", TrustPinConfig{CA: map[string]string{"notary-signer": bundleWithExpiredCertPath}, DisableTOFU: true})
+	require.NoError(t, err)
+
+	certStore.RemoveAll()
+	testPubKey2, err := cryptoService.Create("root", "notary-signer", data.ECDSAKey)
+	require.NoError(t, err)
+	testPrivKey2, _, err := memKeyStore.GetKey(testPubKey2.ID())
+	require.NoError(t, err)
+	expiredCert2, err := cryptoservice.GenerateExpiredTestingCertificate(testPrivKey2.CryptoSigner(), "notary-signer")
+	require.NoError(t, err)
+	allExpiredCertBundle, err := trustmanager.CertChainToPEM([]*x509.Certificate{expiredCert, expiredCert2})
+	require.NoError(t, err)
+	allExpiredCertPath := filepath.Join(tempBaseDir, "all_expired_cert.pem")
+	require.NoError(t, ioutil.WriteFile(allExpiredCertPath, allExpiredCertBundle, 0644))
+	// Now only use expired certs in the bundle, we should fail
+	err = ValidateRoot(certStore, newTestSignedRoot, "notary-signer", TrustPinConfig{CA: map[string]string{"notary-signer": allExpiredCertPath}, DisableTOFU: true})
+	require.Error(t, err)
+
+	// Add an CA for a different GUN that won't validate
```

Comment:
Non-blocking - the GUN seems to be the same as the GUN for `newTestSignedRoot` - it's not using the `goodRootCABundle` - should this comment be just the wrong CA for the key?  (As opposed to one with a different GUN?)